lib/list_versions.go

```diff
@@ -1,142 +1,166 @@
-		//TODO: use GetTFList() instead of GetTFURLBody
-		versions, error := GetTFURLBody(mirrorURL)
-		if error != nil {
-			return "", error
-		}
-		// Getting versions from body; should return match /X.X.X-@/ where X is a number,@ is a word character between a-z or A-Z
-		semver := fmt.Sprintf(`\/(%s{1}\.\d+\-[a-zA-z]+\d*)\/?"`, version)
-		r, err := regexp.Compile(semver)
-		if err != nil {
-			return "", err
-		}
-		for i := range versions {
-			if r.MatchString(versions[i]) {
-				str := r.FindString(versions[i])
-				trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-				return trimstr, nil
-			}
-		}
-	} else if preRelease == false {
-		listAll := false
-		tflist, _ := GetTFList(mirrorURL, listAll) //get list of versions
-		version = fmt.Sprintf("~> %v", version)
-		semv, err := SemVerParser(&version, tflist)
+//GetTFReleases :  Get all terraform releases given the hashicorp url
+func GetTFReleases(mirrorURL string, preRelease bool) ([]*Release, error) {
+	limit := 20
+	queryParams := map[string]string{"limit": strconv.Itoa(limit)}
+	releaseSet, err := getReleases(mirrorURL, queryParams)
+	if err != nil {
+		return nil, err
+	}
+	var releases []*Release
+	releases = append(releases, releaseSet...)
+	for len(releaseSet) == limit {
+		queryParams["after"] = releaseSet[len(releaseSet)-1].TimestampCreated.String()
+		releaseSet, err = getReleases(mirrorURL, queryParams)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		return semv, nil
+		releases = append(releases, releaseSet...)
 	}
-	return "", nil
-}
-
-//GetTFURLBody : Get list of terraform versions from hashicorp releases
-func GetTFURLBody(mirrorURL string) ([]string, error) {
 
-	hasSlash := strings.HasSuffix(mirrorURL, "/")
-	if !hasSlash { //if does not have slash - append slash
-		mirrorURL = fmt.Sprintf("%s/", mirrorURL)
+	if !preRelease {
+		releases = removePreReleases(releases)
 	}
-	resp, errURL := http.Get(mirrorURL)
+	sort.Slice(releases, func(i, j int) bool {
+		return releases[i].Version > releases[j].Version
+	})
+	return releases, nil
+}
+
+//GetTFRelease :  Get the requested terraform release given the hashicorp url
+func GetTFRelease(mirrorURL, requestedVersion string) (*Release, error) {
+	resp, errURL := httpGet(mirrorURL+"/"+requestedVersion, nil)
 	if errURL != nil {
-		log.Printf("[Error] : Getting url: %v", errURL)
-		os.Exit(1)
-		return nil, errURL
+		return nil, fmt.Errorf("[Error] : Getting url: %v", errURL)
 	}
 	defer resp.Body.Close()
 
-	if resp.StatusCode != 200 {
-		log.Printf("[Error] : Retrieving contents from url: %s", mirrorURL)
-		os.Exit(1)
+	body := new(bytes.Buffer)
+	if _, err := io.Copy(body, resp.Body); err != nil {
+		return nil, err
 	}
-
-	body, errBody := ioutil.ReadAll(resp.Body)
-	if errBody != nil {
-		log.Printf("[Error] : reading body: %v", errBody)
-		os.Exit(1)
-		return nil, errBody
+	var release *Release
+	if err := json.Unmarshal(body.Bytes(), &release); err != nil {
```

Comment:
Do we really need to dereference `release` here? It's already a pointer.
Seems like it's unnecessary, but functional:
https://go.dev/play/p/wVqmov9406V

Reply:
I was doing this so I could later return it as a nil'able pointer value. I'd be open to changing it, just don't quite understand what you'd be suggesting.

Reply:
Since we don't need to dereference this, we shouldn't (to keep it simpler)
```suggestion
	if err := json.Unmarshal(body.Bytes(), release); err != nil {
```

The playground link was just to display that these are equivalent.